[MRG] Reduce default learning rate for tSNE to 200.0

iosonofabio opened the pull request:
#### Reference Issue
Fixes #6204 and may alleviate the need for #7215.

#### What does this implement/fix? Explain your changes.
- Changed the default learning rate of tSNE to 200.0 (from 1000.0)
- Added a detailed description in the docstring of how artifacts of too high/low learning rates may look like.

#### Any other comments?
Note: 200 is the standard learning rate of the official tSNE as well:

https://github.com/dominiek/python-bhtsne/blob/master/src/tsne.cpp#L72
https://github.com/dominiek/python-bhtsne/blob/master/src/tsne.cpp#L158

iosonofabio committed:
Reduce default learning rate for tSNE to 200.0

iosonofabio committed:
Add default to class __init__

lesteve commented:
cc @AlexanderFabisch who is the t-SNE expert. In general I am not 100% sure about scikit-learn policy with respect to default parameter changes. Should we have something like defaulting to None and issue a warnings in 0.19 if the parameter is left to the default?

Travis error are genuine, you need to fix the doctest:
```
======================================================================

FAIL: Doctest: sklearn.manifold.t_sne.TSNE

----------------------------------------------------------------------

Traceback (most recent call last):

  File "/home/travis/miniconda/envs/testenv/lib/python3.6/doctest.py", line 2199, in runTest

    raise self.failureException(self.format_failure(new.getvalue()))

AssertionError: Failed doctest test for sklearn.manifold.t_sne.TSNE

  File "/home/travis/build/scikit-learn/scikit-learn/sklearn/manifold/t_sne.py", line 501, in TSNE

----------------------------------------------------------------------

File "/home/travis/build/scikit-learn/scikit-learn/sklearn/manifold/t_sne.py", line 635, in sklearn.manifold.t_sne.TSNE

Failed example:

    model.fit_transform(X) # doctest: +ELLIPSIS, +NORMALIZE_WHITESPACE

Expected:

    array([[ 0.00017599,  0.00003993],

           [ 0.00009891,  0.00021913],

           [ 0.00018554, -0.00009357],

           [ 0.00009528, -0.00001407]])

Got:

    array([[ 0.00017634,  0.0000401 ],

           [ 0.00009807,  0.00022288],

           [ 0.00018649, -0.00009689],

           [ 0.00009507, -0.00001492]])
``

jnothman commented:
yes, that's the idea

On 20 Feb 2017 7:33 pm, "Loïc Estève" <notifications@github.com> wrote:

> cc @AlexanderFabisch <https://github.com/AlexanderFabisch> who is the
> t-SNE expert. In general I am not 100% sure about scikit-learn policy with
> respect to default parameter changes. Should we have something like
> defaulting to None and issue a warnings in 0.19 if the parameter is left to
> the default?
>
> Travis error are genuine, you need to fix the doctest:
>
> ======================================================================
>
> FAIL: Doctest: sklearn.manifold.t_sne.TSNE
>
> ----------------------------------------------------------------------
>
> Traceback (most recent call last):
>
>   File "/home/travis/miniconda/envs/testenv/lib/python3.6/doctest.py", line 2199, in runTest
>
>     raise self.failureException(self.format_failure(new.getvalue()))
>
> AssertionError: Failed doctest test for sklearn.manifold.t_sne.TSNE
>
>   File "/home/travis/build/scikit-learn/scikit-learn/sklearn/manifold/t_sne.py", line 501, in TSNE
>
> ----------------------------------------------------------------------
>
> File "/home/travis/build/scikit-learn/scikit-learn/sklearn/manifold/t_sne.py", line 635, in sklearn.manifold.t_sne.TSNE
>
> Failed example:
>
>     model.fit_transform(X) # doctest: +ELLIPSIS, +NORMALIZE_WHITESPACE
>
> Expected:
>
>     array([[ 0.00017599,  0.00003993],
>
>            [ 0.00009891,  0.00021913],
>
>            [ 0.00018554, -0.00009357],
>
>            [ 0.00009528, -0.00001407]])
>
> Got:
>
>     array([[ 0.00017634,  0.0000401 ],
>
>            [ 0.00009807,  0.00022288],
>
>            [ 0.00018649, -0.00009689],
>
>            [ 0.00009507, -0.00001492]])
> ``
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/8397#issuecomment-281017035>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz6zdXwXOVX8kXdvDXor9AxRVicdmcks5reU_RgaJpZM4MFPdL>
> .
>

lesteve commented:
@iosonofabio have you tried running the script in https://github.com/jjvalletta/t-SNEIssue to see if changing the default parameter value fixes the original problem?

iosonofabio commented:
Thx for the quick feedback, will fix the doctest and test that script today.

I think 200 should be the default since that is that the original C++ tSNE code does, shouldn't it?

On February 20, 2017 4:25:32 AM PST, "Loïc Estève" <notifications@github.com> wrote:
>@iosonofabio have you tried running the script in
>https://github.com/jjvalletta/t-SNEIssue to see if changing the default
>parameter value fixes the original problem?
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/scikit-learn/scikit-learn/pull/8397#issuecomment-281066325

lesteve commented:
> I think 200 should be the default since that is that the original C++ tSNE code does, shouldn't it?

This is a reason to change the default parameter but I don't think it is enough by itself. We need to make sure we understand the implications of changing the default parameter.

Amongst other things it would be good to check that the problem mentioned in the original issue is fixed by this change. If you have a simpler stand-alone snippet reproducing the problem even better. A test would be great too.

Also even if everyone agrees that changing the default parameter value is the right way forward, we still need to think about how to warn the user that something has changed (one possible mechanism is mentioned in https://github.com/scikit-learn/scikit-learn/pull/8397#issuecomment-281017035).

iosonofabio committed:
Change doctest to match new default learning rate

iosonofabio commented:
- I changed the failing doctest, it should work now.
- I can confirm that reducing the default rate to 200 fixes the original issue: https://github.com/iosonofabio/t-SNEIssue (see `script.py` in that repo for a snippet that fixes the issue)
- Another example is given by the [alternative python module for tSNE](https://github.com/dominiek/python-bhtsne) on the IRIS data set. This example is also fixed by the change. It is reported in the [same repo](https://github.com/iosonofabio/t-SNEIssue) below the first example (see `comparison.py` for the snippet).
- I would advise against setting the default rate to `None` and forcing the user to explicitely select a rate. The learning rate is an implementation detail of the adaptive version of gradient descent. Although it does affect results if set grossly wrong (because we get stuck in the wrong local minimum of the optimization function), we _do_ have a working default and that is valuable. In the original code and in the `bhtsne` version of the algorithm, the rate cannot even be changed, yet that code has been used in serious research without problems, e.g. in the [wishbone module](https://github.com/ManuSetty/wishbone) and its associated [Nature Biotechnology paper](http://www.nature.com.ezproxy.stanford.edu/nbt/journal/v34/n6/full/nbt.3569.html).
- Instead, I propose we just include the change in the release notes.

lesteve commented:
> I would advise against setting the default rate to None and forcing the user to explicitely select a rate.

My proposal was slightly different so I am going to make it more explicit:
* default to None as a way to figure out whether the user is using the default value
* if the value is set to None: warn that the default value has changed in scikit-learn 0.19 from 1000 to 200.
* if the value is not set to None then no warnings should be issued (this should be suggested as a way to get rid of the warning in the warning message)

> Instead, I propose we just include the change in the release notes.

This will be necessary (i.e. you will need to add an entry in doc/whats_new.rst) but I don't feel this is enough. Silent changes have a very negative impact on users and let's be honest we can not expect all the users to read the release notes prior to updating.

lesteve commented:
> I can confirm that reducing the default rate to 200 fixes the original issue: https://github.com/iosonofabio/t-SNEIssue (see script.py in that repo for a snippet that fixes the issue)
Another example is given by the alternative python module for tSNE on the IRIS data set. This example is also fixed by the change. It is reported in the same repo below the first example (see comparison.py for the snippet).

Thanks a lot for this! Could you post plots you obtained for these two scripts (for both default values). This will be very useful to add some context if we need to revisit the default value choice in the future.

iosonofabio commented:
Hi @lesteve,

- the plots from those two scripts are already shown at https://github.com/iosonofabio/t-SNEIssue, or is that not what you meant?
- I understand your concerns, on the other hand function/class signatures matter. Setting a default value of `None` but defaulting to 200 sounds like an antipattern to me. Anyone checking the function signature (e.g. `TSNE?` in `ipython`) will get confused. I would find it fine, if a bit verbose, to print a deprecation warning to all users that require `learning_rate=200`, until `sklearn 0.20` is out or so.

Just to put things a bit in perspective, in general I agree that API changes require warnings and silent changes are heinous. However, in this particular case we just got it wrong the first time and we are repairing. The current default is completely broken as I can tell from real-life examples, moreover it does not reflect the code of the tSNE author. Older versions of `sklearn` did the right thing, I assume because they did not include the Barnes-Hut approximation, and at some point during that implementation (I guess around beginning 2016) things went awry and the default settings stopped giving usable results. People are scared enough of `sklearn`'s default that they end up relying on alternative implementations altogether (e.g. [bhtsne](https://github.com/dominiek/python-bhtsne)) because they cannot be bothered learning all bells and whistles of our settings while the simpler implementations just do the right thing. So in a nutshell, silence is bad but hiding secret defaults behind `None`s and forcing the user to mess around with learning rates might disappoint even more people...

jnothman commented:
If you don't like learning_rate=None, you could use
learning_rate='1000-deprecated'....?

On 21 February 2017 at 10:57, Fabio Zanini <notifications@github.com> wrote:

> Hi @lesteve <https://github.com/lesteve>,
>
>    - the plots from those two scripts are already shown at
>    https://github.com/iosonofabio/t-SNEIssue
>    <https://github.com/iosonofabio/t-SNEIssue>, or is that not what you
>    meant?
>    - I understand your concerns, on the other hand function/class
>    signatures matter. Setting a default value of None but defaulting to
>    200 sounds like an antipattern to me. Anyone checking the function
>    signature (e.g. TSNE? in ipython) will get confused. I would find it
>    fine, if a bit verbose, to print a deprecation warning to all users that
>    require learning_rate=200, until sklearn 0.20 is out or so.
>
> Just to put things a bit in perspective, in general I agree that API
> changes require warnings and silent changes are heinous. However, in this
> particular case we just got it wrong the first time and we are repairing.
> The current default is completely broken as I can tell from real-life
> examples, moreover it does not reflect the code of the tSNE author. Older
> versions of sklearn did the right thing, I assume because they did not
> include the Barnes-Hut approximation, and at some point during that
> implementation (I guess around beginning 2016) things went awry and the
> default settings stopped giving usable results. People are scared enough of
> sklearn's default that they end up relying on alternative implementations
> altogether (e.g. bhtsne <https://github.com/dominiek/python-bhtsne>)
> because they cannot be bothered learning all bells and whistles of our
> settings while the simpler implementations just do the right thing. So in a
> nutshell, silence is bad but hiding secret defaults behind Nones and
> forcing the user to mess around with learning rates might disappoint even
> more people...
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/8397#issuecomment-281208754>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz691ejDjYETjqZ4BqsrYj358VSRe4ks5reihggaJpZM4MFPdL>
> .
>

jnothman commented:
The point is that if someone explicitly sets learning_rate=1000 which they
might do in order to maintain the behaviour they experienced on the
previous scikit-learn release, they shouldn't be forced to see a
deprecationwarning.

On 21 February 2017 at 11:22, Joel Nothman <joel.nothman@gmail.com> wrote:

> If you don't like learning_rate=None, you could use learning_rate='1000-
> deprecated'....?
>
> On 21 February 2017 at 10:57, Fabio Zanini <notifications@github.com>
> wrote:
>
>> Hi @lesteve <https://github.com/lesteve>,
>>
>>    - the plots from those two scripts are already shown at
>>    https://github.com/iosonofabio/t-SNEIssue
>>    <https://github.com/iosonofabio/t-SNEIssue>, or is that not what you
>>    meant?
>>    - I understand your concerns, on the other hand function/class
>>    signatures matter. Setting a default value of None but defaulting to
>>    200 sounds like an antipattern to me. Anyone checking the function
>>    signature (e.g. TSNE? in ipython) will get confused. I would find it
>>    fine, if a bit verbose, to print a deprecation warning to all users that
>>    require learning_rate=200, until sklearn 0.20 is out or so.
>>
>> Just to put things a bit in perspective, in general I agree that API
>> changes require warnings and silent changes are heinous. However, in this
>> particular case we just got it wrong the first time and we are repairing.
>> The current default is completely broken as I can tell from real-life
>> examples, moreover it does not reflect the code of the tSNE author. Older
>> versions of sklearn did the right thing, I assume because they did not
>> include the Barnes-Hut approximation, and at some point during that
>> implementation (I guess around beginning 2016) things went awry and the
>> default settings stopped giving usable results. People are scared enough of
>> sklearn's default that they end up relying on alternative
>> implementations altogether (e.g. bhtsne
>> <https://github.com/dominiek/python-bhtsne>) because they cannot be
>> bothered learning all bells and whistles of our settings while the simpler
>> implementations just do the right thing. So in a nutshell, silence is bad
>> but hiding secret defaults behind Nones and forcing the user to mess
>> around with learning rates might disappoint even more people...
>>
>> —
>> You are receiving this because you commented.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/scikit-learn/scikit-learn/pull/8397#issuecomment-281208754>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AAEz691ejDjYETjqZ4BqsrYj358VSRe4ks5reihggaJpZM4MFPdL>
>> .
>>
>
>

AlexanderFabisch commented:
By the way: there are the following examples that work with the default learning rate: [1](http://scikit-learn.org/stable/auto_examples/manifold/plot_compare_methods.html#sphx-glr-auto-examples-manifold-plot-compare-methods-py), [2](http://scikit-learn.org/stable/auto_examples/manifold/plot_lle_digits.html#sphx-glr-auto-examples-manifold-plot-lle-digits-py), [3](http://scikit-learn.org/stable/auto_examples/manifold/plot_manifold_sphere.html#sphx-glr-auto-examples-manifold-plot-manifold-sphere-py). You either would have to set 1000 explicitly or you have to check if they work with 200.

iosonofabio commented:
I tested all three with both `learining_rate=1000` and `200`. All three give very similar results, 1.5-3X faster with the higher learning rate (as expected). So we lose nothing but a bit of speed by reducing the default, which is a best-case scenario: people that want to risk landing in a worse minimum for speed reasons should feel free to increase the learning rate.

In exchange, we gain correct behaviour on real datasets (e.g. single-cell transcriptome data).

lesteve commented:
> the plots from those two scripts are already shown at https://github.com/iosonofabio/t-SNEIssue, or is that not what you meant?

OK I missed that, it is still useful to include the images in the description so we have a snapshot in the discussion in case something changes in your repo.

> By the way: there are the following examples that work with the default learning rate: 1, 2, 3. You either would have to set 1000 explicitly or you have to check if they work with 200. Could you check that?

Good thinking @AlexanderFabisch I will push a change in this branch that will force building the doc on CircleCI, this way we can compare the output of the examples for this PR vs master.

lesteve committed:
[doc build] empty commit to generate the doc

lesteve commented:
Here are the PR vs master links/plots to make the comparison easier:

### plot_compare_methods:
[master](http://scikit-learn.org/dev/auto_examples/manifold/plot_compare_methods.html)
![](http://scikit-learn.org/dev/_images/sphx_glr_plot_manifold_sphere_001.png)

[PR](https://9005-843222-gh.circle-artifacts.com/0/home/ubuntu/scikit-learn/doc/_build/html/stable/auto_examples/manifold/plot_compare_methods.html)
![](https://9005-843222-gh.circle-artifacts.com/0/home/ubuntu/scikit-learn/doc/_build/html/stable/_images/sphx_glr_plot_manifold_sphere_001.png)

### plot_lle_digits:
[master](http://scikit-learn.org/dev/auto_examples/manifold/plot_lle_digits.html#sphx-glr-auto-examples-manifold-plot-lle-digits-py)
![](http://scikit-learn.org/dev/_images/sphx_glr_plot_lle_digits_013.png)

[PR](https://9005-843222-gh.circle-artifacts.com/0/home/ubuntu/scikit-learn/doc/_build/html/stable/auto_examples/manifold/plot_lle_digits.html)
![](https://9005-843222-gh.circle-artifacts.com/0/home/ubuntu/scikit-learn/doc/_build/html/stable/_images/sphx_glr_plot_lle_digits_013.png)

### plot_manifold_sphere:
[master](http://scikit-learn.org/dev/auto_examples/manifold/plot_manifold_sphere.html)
![](http://scikit-learn.org/dev/_images/sphx_glr_plot_compare_methods_001.png)

[PR](https://9005-843222-gh.circle-artifacts.com/0/home/ubuntu/scikit-learn/doc/_build/html/stable/auto_examples/manifold/plot_manifold_sphere.html)
![](https://9005-843222-gh.circle-artifacts.com/0/home/ubuntu/scikit-learn/doc/_build/html/stable/_images/sphx_glr_plot_compare_methods_001.png)

iosonofabio commented:
Thanks @lesteve, I'll try to post the other images from my repo too (I have never attached images to GH posts). The names of the scripts in the last post are shuffled but they look good!

lesteve commented:
> The names of the scripts in the last post are shuffled but they look good!

Good point it should be fixed now.

iosonofabio committed:
Mention in whats_new.rst

lesteve commented:
@AlexanderFabisch can you have a quick look at the plots above [here](https://github.com/scikit-learn/scikit-learn/pull/8397#issuecomment-281287888) and [here](https://github.com/scikit-learn/scikit-learn/pull/8397#issuecomment-281406580)

Looking just at the timings from the examples, it looks like learning_rate=200 is 2-3x times slower than with learning_rate=1000. Not sure how much of a concern this is.

@iosonofabio we will still need the warning messages as mentioned above.

AlexanderFabisch commented:
The result of plot_manifold_sphere looks even better to me. We could set a higher learning rate in the other examples if we want that. I'm not sure what to do.

iosonofabio committed:
Warn user if learning rate of tSNE is default

lesteve commented on sklearn/manifold/t_sne.py:
```diff
@@ -672,6 +676,12 @@ def __init__(self, n_components=2, perplexity=30.0,
         self.angle = angle
 
+        # TODO sklearn 0.20: remove this warning
+        if learning_rate == 200.0:
```
Please read again https://github.com/scikit-learn/scikit-learn/pull/8397#issuecomment-281136788 and try to follow what I suggested.

If you don't like setting the default to None, use something else like `'default-changed-from-1000-to-200-in-0.19'` or something better.

iosonofabio replied:
As mentioned above I feel that makes things worse because:
- it obscures the actual default, which will be 200
- it confuses the user, because `learning_rate` is a float but gets defaulted to a string
- it worsers the function signature

I think there is nothing wrong in warning everybody for a development cycle, a year from now we won't even be talking about this anymore.

AlexanderFabisch replied:
People will probably find strange ways to turn this warning off. Maybe that means setting `learning_rate=199.99` which is certainly worse.

lesteve replied:
Let's try to find consensus on the default value change before we decide on how we want the warnings to happen. See https://github.com/scikit-learn/scikit-learn/pull/8397#issuecomment-281628367.

iosonofabio replied:
there does not seem to be a consensus reached, nonetheless I now set it to None hoping this increases the chances the PR gets anywhere.

lesteve commented:
It feels like having a default to 200 is the right thing to do even if there is a 2x-3x slowdown. Having a default value for which the algorithm does not converge is less than great. It would be interesting to have others opinion to make sure there is consensus on this @jnothman @ogrisel @GaelVaroquaux @amueller @agramfort.

In case there is consensus, I think what is left to do is how to handle the change of the default value. Here are a few ways I can see:

* change the default to 200 and documenting prominently the change in the changelog. The easiest in term of maintenance but the most surprising to the users that probably do not read the changelog before upgrading.

* learning_rate defaults to a sentinel value and warn only if `learning_rate==sentinel_value` for 2 releases. Getting rid of the warning is possible by setting learning_rate explicitly (maybe we can find a better mechanism of disabling the warning). Easy in term of maintenance. Encouraging users to set an explicit learning_rate may not be that great since it can be considered as an implementation detail that most users do not want to know about.

* suggestion from @ogrisel: FutureWarning that the default will change in a future release. Leave the default to 1000 for 1 or 2 releases and then change it. Probably the cleanest way in term of not surprising users but it will take a while to reach users, which is not so great if we think 200 is the right default.

As usual this is a trade-off trying to balance many things among which user experience and cost of maintenance.

iosonofabio commented:
Thanks for the nice summary of possibilities. Will leave space for others to comment, I'm just a bit surprised that we are so afraid. People are using an algorithm that does not converge in most cases and will suddenly have better results...

As opposed to we carry on very cautiously and the same people publish papers with bogus dimensional reductions and cite `sklearn` for that.

The balance seems obvious to me.

kylemcdonald commented:
Just wanted to chime in and say: as a frequent t-SNE user and advocate, I guide people away from the sklearn implementation primarily because the learning rate is set so high (and a sometimes because it's slower, but that doesn't really matter for small datasets).

For what it's worth, the [original paper (2008)](https://lvdmaaten.github.io/publications/papers/JMLR_2008.pdf) set the learning rate to 100, and uses an adaptive learning rate based on the gains. If you look at the [Matlab implementation](https://lvdmaaten.github.io/tsne/code/tSNE_matlab.zip) it is initially set to 500. In the updated paper describing [Barnes-Hut t-SNE (2014)](https://lvdmaaten.github.io/publications/papers/JMLR_2014.pdf) he doesn't provide a learning rate, but the "official" implementation of bhtsne from the author [sets it to 200](https://github.com/lvdmaaten/bhtsne/blob/master/tsne.cpp#L72) with the same gains-adapting technique as before.

In short, this PR seems pretty clearly to be going in the right direction. It would be great to see it merged.

AlexanderFabisch commented:
I would vote for option 2 (sentinel value, [example](https://github.com/AlexanderFabisch/scikit-learn/commit/cfdece6343517dd08094377a0171103a0d3a82ca)). I don't see any problem here. The learning rate is a critical parameter that I would set manually anyway in most cases.

lesteve commented:
> I would vote for option 2 (sentinel value, example).

I am leaning towards this as well and the people I asked IRL were split between option 1 and 2. Let's go for option 2 then. @iosonofabio can you implement option 2?

iosonofabio commented:
I can do many things but am strongly in favour of 1. If the learning rate is a critical parameter we are doing it wrong: this number cannot even be set in the official t-SNE implementation and I never heard anybody complaining.

On March 14, 2017 3:23:05 AM PDT, "Loïc Estève" <notifications@github.com> wrote:
>> I would vote for option 2 (sentinel value, example).
>
>I am leaning towards this as well and the people I asked IRL were split
>between option 1 and 2. Let's go for option 2 then. @iosonofabio can
>you implement option 2?
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/scikit-learn/scikit-learn/pull/8397#issuecomment-286379687

lesteve commented:
> I can do many things but am strongly in favour of 1

The consensus seems to be leaning towards option 2. so it would be great if you could implement it.

jmschrei commented:
I'm in favor of the second option. Particularly with t-SNE you seem to have to tune a lot of parameters yourself anyway.

lesteve commented:
It would be great to finish up this PR, @iosonofabio do you have the spare bandwith to do it?

iosonofabio committed:
Default learning_rate=None

iosonofabio commented:
I think setting `learning_rate=None` is an antipattern and did not see a clear consensus on it, but it seems to be the only compromise that will ever be merged, so I changed it that way.

iosonofabio committed:
Merge remote-tracking branch 'sklearn/master'

iosonofabio committed:
Merge whats_new

iosonofabio committed:
Forgot self. in setting tSNE learning rate

iosonofabio commented:
Travis fails with the new defaulting to `None`. We have a default of `None` in the constructor but then we set it to `200`: now `sklearn.utils.estimator_checks.check_parameters_default_constructible` thinks we are using bad coding style (via the `inspect` module).

I have no understanding of the ins and outs of unit testing in scikit-learn, but the machine is basically right that faking the signature of `TSNE.__init__` is essentially a bad idea. This is why I advocate again for just setting the default to `200.0` and adding a note in the release history.

jnothman commented on sklearn/manifold/t_sne.py:
```diff
@@ -674,6 +678,14 @@ def __init__(self, n_components=2, perplexity=30.0,
+            warnings.warn("The default learning rate of TSNE has changed " +
+                          "from 1000 to 200 in sklearn 0.19. Set the " +
+                          "learning rate explicitely to avoid this message.")
+            self.learning_rate = 200.0
```
The only problem is that you're doing this in `__init__` when you should be doing it in `fit`.

iosonofabio replied:
I doubt `fit` is the right place since it just cals `fit_transform`. Maybe you would like `fit_transform`? But then again, that only calls `_fit`. Maybe `_fit`? Ouch, that only calls `_tsne`. So maybe we got it, it must be in `_tsne`... or not? Because actually `_tsne` only calls `_gradient_descent`....??

Now it *would* work if I changed it in `_gradient_descent`, but `_gradient_descent`:
- is very deep in the hierarchy
- relies heavily on proper instance attributes: it is a numerical recipe, not a playground for object-oriented interfaces

The only way to fix it using "solution 2" while keeping unit tests happy is by breaking even more function signatures and messing around with the deepest numerical function in the code. Or maybe that's going to break more tests? Or maybe some tests in the future? Because we all know burying such fake interfaces deep into the code is bad design.

Bottomline, this argues again for "solution 1", which is just change the constructor default to 200 and tell users about it in the release notes and, if you so wish, upon module import.

ogrisel commented:
Closing in favor of #9032 that sets the default learning rate to 200 along with many other fixes in the optimizer scheduling.
